### ISSUE TITLE
Dynamic camera, first pass

### ISSUE THREAD

**Issue by ericrwolfe** [mapbox/mapbox-navigation-ios#773]
The camera should:
* Be more dynamic, communicating maneuver urgency by camera distance/zoom level
* Be timed with voice announcements: what's communicated on screen should match audio cues
* Provide ample context for maneuvers

As a rough first pass, we should implement basic stepped camera movements achieving the following:
- [x] Triggered by/synchronized with voice announcements
- [x] On the completion announcement of a maneuver: zoom out to provide context for long steps with long periods of inactivity (https://github.com/mapbox/mapbox-navigation-ios/pull/769)
- [ ] At the first approach maneuver announcement (`In 2 miles...`): zoom to show the upcoming maneuver on the map, whether by a fixed zoom level based on distance, or a zoom to fit function.
- [x] At the final turn maneuver announcement: zoom in closely to our current default zoom level 

Once this basic set of triggers/logic is in place, we can experiment with more sophisticated zoom logic, camera tilt, anchor points, etc. @d-prukop

cc @jthramer @bsudekum @danesfeder

**Comment by JThramer** [comment]
@ericrwolfe A question about Item 3: Do you want to set the altitude _once_ at maneuver announcement time? or do you want to continuously zoom in as the user gets closer to the upcoming maneuver, a la Lyft/Uber? Both are doable, but the latter might be a little bit tricker, as there may be some `userTrackingMode` implications, as well as the need to hook into `NavigationController.progressDidChange(notification:)`.

/cc @bsudekum @frederoni @1ec5

**Comment by ericrwolfe** [comment]
> Do you want to set the altitude once at maneuver announcement time? or do you want to continuously zoom in as the user gets closer to the upcoming maneuver

@JThramer For this first pass we'll just zoom in once at each step. We'll explore continuous zoom in follow on work. cc @d-prukop

**Comment by JThramer** [comment]
Closing as the development cycle seems to be over for this work.

**Comment by ericrwolfe** [comment]
@JThramer did we ever implement the approach maneuver announcement zooming to show the upcoming maneuver (item 3)? If not, can you open another ticket as tail work with some background on any approaches you explored?

**Comment by JThramer** [comment]
We did not. IIRC this was because of a technical limitation -- we currently do not have a good mechanism to "zoom-to-fit" the camera in the context we were looking for ("Given our current camera position, what camera altitude would keep these N points in-view?"), and lookup-tables based on distance proved to be problematic due to reasons that I think were related to speed. We ended up implementing an incremental change in #781, where the camera "zooms in" to the "default altitude" (1000m) at the last instruction of a step. I'll create a follow-up ticket for item 3.

**Comment by 1ec5** [comment]
> we currently do not have a good mechanism to "zoom-to-fit" the camera in the context we were looking for ("Given our current camera position, what camera altitude would keep these N points in-view?")

There is such a mechanism, which we use in overview mode:

https://github.com/mapbox/mapbox-navigation-ios/blob/49d8d3e9014bd97fb6fadee7703cfae888d003a3/MapboxNavigation/RouteMapViewController.swift#L281-L298

`MGLMapView.setVisibleCoordinateBounds(_:edgePadding:animated:)` and `setVisibleCoordinates(_:count:edgePadding:animated:)` do make it possible to keep an array of points in view. For a more sophisticated effect, check out MGLMapView’s [`targetCoordinate`](https://www.mapbox.com/ios-sdk/api/3.7.1/Classes/MGLMapView.html#/c:objc%28cs%29MGLMapView%28py%29targetCoordinate). Though it only works in conjunction with MGLMapView’s built-in course tracking mode, which RouteMapViewController no longer uses, [the same implementation](https://github.com/mapbox/mapbox-gl-native/blob/f4021c4a2ee985e33bdd088985a22c591503d234/platform/ios/src/MGLMapView.mm#L5019-L5054) could be ported to the navigation SDK without much fuss.

**Comment by d-prukop** [comment]
@JThramer, @asheemmamoowala built a method on the SDK that may help you out here. Check out https://github.com/mapbox/mapbox-gl-native/issues/9809

**Comment by 1ec5** [comment]
Yes, `-[MGLMapView cameraThatFitsShape:direction:padding:]` is potentially even better than just passing two coordinates into `setVisibleCoordinates(_:count:edgePadding:animated:)`, because it can fit to a very windy step without letting any of the step fall off the edges of the viewport.
